### PR TITLE
Refactor exception handling

### PR DESCRIPTION
**Pull request by pnwpedro**
This moves a lot of things around, and I still have a lot of tests to write.

I think this will be easier to manage going forward, though.

### CODE REVIEW

File: Fauna/Exceptions/ExceptionFactory.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            "invalid_request" => new InvalidRequestException(msg, f),
+            "limit_exceeded" => new ThrottlingException(msg, f),
+            "time_limit_exceeded" => new QueryTimeoutException(msg, f),
+            "timeout" or
+                "time_out" or
+                "gateway_timeout" => new TimeoutException(msg, f),
+            "unauthorized" => new UnauthorizedException(msg, f),
+
+            _ => new ServiceException(msg, f)
+        };
```
Comment (adambollen) [inline Fauna/Exceptions/ExceptionFactory.cs:33]: Nit: for readability:
```suggestion
        return f.ErrorCode switch
        {
            "abort"                 => new AbortException(msg, f, ctx),
            "bad_gateway"           => new BadGatewayException(msg, f),
            "contended_transaction" => new ContendedTransactionException(msg, f),
            "forbidden"             => new ForbiddenException(msg, f),
            "invalid_argument"      => new QueryRuntimeException(msg, f),
            "invalid_query" or
                "invalid_function_definition" or
                "invalid_identifier" or
                "invalid_syntax" or
                "invalid_type"      => new QueryCheckException(msg, f),
            "invalid_request"       => new InvalidRequestException(msg, f),
            "limit_exceeded"        => new ThrottlingException(msg, f),
            "time_limit_exceeded"   => new QueryTimeoutException(msg, f),
            "timeout" or
                "time_out" or
                "gateway_timeout"   => new TimeoutException(msg, f),
            "unauthorized"          => new UnauthorizedException(msg, f),

            _ => new ServiceException(msg, f)
        };
```

File: Fauna/Exceptions/ExceptionFactory.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            return new ProtocolException("Malformed response.", r.StatusCode, body);
+        }
+
+        return r.StatusCode switch
+        {
+            HttpStatusCode.TooManyRequests => new ThrottlingException($"{r.StatusCode}: {r.ReasonPhrase ?? "Too many requests."}"),
+            _ => new FaunaException($"{r.StatusCode}: {r.ReasonPhrase}")
+        };
+    }
+}
```
Comment (adambollen) [inline Fauna/Exceptions/ExceptionFactory.cs:51]: Nit: newline
```suggestion
}

```

File: Fauna/Response/QueryResponse.cs [hunk 3]
```diff
@@ -1,3 +1,4 @@
+using System.Net;
 using Fauna.Mapping;
 using Fauna.Serialization;
 using System.Text.Json;
```
Comment (adambollen) [inline Fauna/Response/QueryResponse.cs:4]: Nit: Put all the `System.*` usings at the top.
```suggestion
using System.Net;
using System.Text.Json;
using Fauna.Mapping;
using Fauna.Serialization;
```

File: Fauna/Exceptions/ExceptionFactory.cs (hunk 1 above)
Comment (pnwpedro) [reply Fauna/Exceptions/ExceptionFactory.cs:33]: i'm not sure if the linter will keep this. but i'll commit it

Comment (adambollen) [reply Fauna/Exceptions/ExceptionFactory.cs]: Hehe, true.  Probably didn't like it; oh well.

File: Fauna/Exceptions/ExceptionFactory.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            "invalid_argument" => new QueryRuntimeException(msg, f),
+            "invalid_query" or
+                "invalid_function_definition" or
+                "invalid_identifier" or
+                "invalid_syntax" or
+                "invalid_type" => new QueryCheckException(msg, f),
+            "invalid_request" => new InvalidRequestException(msg, f),
+            "limit_exceeded" => new ThrottlingException(msg, f),
+            "time_limit_exceeded" => new QueryTimeoutException(msg, f),
+            "timeout" or
```
Comment (freels) [inline Fauna/Exceptions/ExceptionFactory.cs:27]: As discussed you can remove the "timeout" variant if you'd like.

**Review by freels — APPROVED** [review]
seems like a significant improvement

File: Fauna/Exceptions/ExceptionFactory.cs (hunk 4 above)
Comment (pnwpedro) [reply Fauna/Exceptions/ExceptionFactory.cs:27]: right, i'll address that separately

**Pull request merged**